pkg/ostree/app.js

```diff
@@ -0,0 +1,328 @@
+            })
+            .fail(function(ex) {
+                scope.$applyAsync(function() {
+                    scope.error = cockpit.format(cockpit.message(ex));
+                });
+            })
+            .always(function(ex) {
+                scope.$applyAsync(function() {
+                    scope.progressMsg = null;
+                });
+            });
+    }
+
+
+    return angular.module('ostree', [
+            'ngRoute',
+        ])
+        .config([
+            '$routeProvider',
+             function($routeProvider) {
+                $routeProvider.when('/', {
+                    templateUrl: 'index.html',
+                    controller: 'indexController'
+                });
+        }])
+
+        .controller('mainController', [
+            '$scope',
+            '$timeout',
+            'ostreeClient',
+            function($scope, $timeout, client) {
+                function show_failure(ex) {
+                    var message = null;
+                    if (ex.problem === "access-denied")
+                        message = _("Not authorized to update software on this system");
+                    else
+                        message = cockpit.message(ex);
+
+                    $scope.curtains = { state: 'failed',
+                                        failure: ex,
+                                        message: message };
+                    $scope.$digest();
+                    phantom_checkpoint();
+                }
+
+                function check_empty() {
+                    if (timeout)
+                        window.clearTimeout(timeout);
+
+                    timeout = null;
+                    if (client.os_list && client.os_list.length === 0) {
+                        $scope.curtains = {
+                            "state" : "empty",
+                            "message" : _("No OSTree deployments found"),
+                            "failure" : true
+                        };
+                        $scope.$digest();
+                        phantom_checkpoint();
+                    } else if ($scope.curtains !== null) {
+                        $scope.curtains = null;
+                        $scope.$digest();
+                        phantom_checkpoint();
+                    }
+                }
+
+                $scope.curtains = { state: 'silent' };
+                var timeout = window.setTimeout(function() {
+                    $scope.curtains = { state: 'connecting' };
+                    $scope.$digest();
+                    timeout = null;
+                }, 1000);
+
+                function handle(promise) {
+                    promise
+                        .always(function() {
+                            window.clearTimeout(timeout);
+                            timeout = null;
+                        })
+                        .done(function(connection) {
+                            $(client).on("connectionLost.main", function(event, ex) {
```

Comment:
For multiple reconnects theseh event handlers will be connected multiple times. Why not just connect them once?